custom/zipline/models.py

```diff
@@ -156,6 +179,166 @@ class EmergencyOrder(models.Model):
         related_name='+', null=True)
 
 
+class EmergencyOrderPackage(models.Model):
+
+    class Meta:
+        index_together = [
+            ['order', 'package_number'],
+        ]
+
+        unique_together = [
+            ['order', 'package_number'],
+        ]
+
+    # (order, package_number) matches up with the same-named fields on EmergencyOrderStatusUpdate
+    order = models.ForeignKey('EmergencyOrder')
+    package_number = models.IntegerField()
+
+    # Same format as EmergencyOrder.products_requested; represents products and quantities
+    # in this package
+    products = jsonfield.JSONField(default=dict)
+
+    # Should either be STATUS_DISPATCHED, STATUS_DELIVERED, or STATUS_CANCELLED_IN_FLIGHT
+    status = models.CharField(max_length=126, default=EmergencyOrderStatusUpdate.STATUS_DISPATCHED)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the dispatched status update
+    # for this package
+    dispatched_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the delivered status update
+    # for this package
+    delivered_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # A pointer to the EmergencyOrderStatusUpdate representing the cancelled in flight status update
+    # for this package
+    cancelled_status = models.ForeignKey('EmergencyOrderStatusUpdate', on_delete=models.PROTECT,
+        related_name='+', null=True)
+
+    # The total cost of all products in this package, using info from OrderableProduct
+    cost = models.DecimalField(max_digits=12, decimal_places=3, null=True)
+
+    # The total weight of this package in grams, using info from OrderableProduct
+    weight = models.DecimalField(max_digits=12, decimal_places=3, null=True)
+
+    def update_calculated_fields(self):
+        """
+        Updates the cost and the weight of the package.
+        Note that this uses the active record for the product, so this method
+        should only be called in real-time (not on historical data).
+        """
+        from custom.zipline.api import log_zipline_exception
+
+        cost = Decimal(0)
+        weight = Decimal(0)
+
+        for code, data in self.products.iteritems():
+            try:
+                product = OrderableProduct.objects.get(domain=self.order.domain, code=code)
+            except OrderableProduct.DoesNotExist:
+                log_zipline_exception(
```

Comment:
yeah, so my thought process here was that technically this should never happen since we do validation earlier on in the workflow on the product code. but if this error does happen here, we don't want to block the workflow from completing by raising the exception, so since all we miss out on here is some reporting data (this isn't info that's critical to the workflow completing), we just log the exception so someone can look into it.